codegen/src/CodegenError.scala

```diff
@@ -0,0 +1,45 @@
+package besom.codegen
+
+@SerialVersionUID(1L)
+sealed abstract class CodegenError(message: Option[String], cause: Option[Throwable])
+    extends Exception(message.orElse(cause.map(_.toString)).orNull, cause.orNull)
+    with Product
+    with Serializable
+
+@SerialVersionUID(1L)
+case class GeneralCodegenException(message: Option[String], cause: Option[Throwable])
+    extends CodegenError(message, cause)
+object GeneralCodegenException {
+  def apply(message: String)                   = new GeneralCodegenException(Some(message), None)
+  def apply(message: String, cause: Throwable) = new GeneralCodegenException(Some(message), Some(cause))
+  def apply(cause: Throwable)                  = new GeneralCodegenException(None, Some(cause))
+  def unapply(e: GeneralCodegenException): Option[(String, Throwable)] = Some((e.getMessage, e.getCause))
```

Comment:
Are these extractors ever used? They seem to assume that message and cause are always defined, which would be contrary to defining them as `Option`s

Reply:
Not at the moment, no. They were added for the sake of completeness.

How should they be changed to be useful then?

Reply:
The auto-generated extractor would work e.g. like this
```scala
case GeneralCodegenException(Some(message), None) =>
  println(s"Extracted message: ${message}. Cause missing")
```
The custom extractor would instead extract message as a string not wrapped into an option and the cause would be a null. I'm not even sure how exactly the compiler would decide which overloaded variant of `unapply` to use so this is likely to cause surprising bugs. Let's just use the default extractor until it turns out we really need something more convenient and we know exactly what

Reply:
right, maybe I'll just let the auto-generated do it's job until a custom is actually needed. Thank you, you were right to point this out as unnecessary :)